### ISSUE TITLE
Too much magic in `Rails.logger` and logging to STDERR causes duplicate output.

### ISSUE THREAD

**Issue by ioquatix** [rails/rails#44800]
With the following code in a Rails configuration file:

```ruby
Rails.logger = Logger.new(STDERR)
```
Adding the following to a rails app configuration:

```ruby
Rails.logger = Logger.new(STDERR)
```

Causes duplicate log statements when running the app with `rails s`. It's caused by the following:

https://github.com/rails/rails/blob/346ae79d5a11152f508df8d5506f8a0682ddb74b/railties/lib/rails/commands/server/server_command.rb#L75-L85

The implementation of `logger_outputs_to?` uses private details of the logger instance to determine whether it's logging to `STDOUT`. In this specific case, adding `STDERR` as part of the check might also work. However, not all logger objects have the same private details causing this check to fail in those cases too.

https://github.com/rails/rails/blob/346ae79d5a11152f508df8d5506f8a0682ddb74b/activesupport/lib/active_support/logger.rb#L16-L20

I don't think we should be inspecting private object details and I believe that the implementation creates too many edge cases which are better handled by explicit configuration.

My advice would be:

- Default to `Rails.logger = Logger.new(STDERR)` (& don't log to log files).
- Don't use `instance_variable_get` and consider removing `logger_outputs_to?`.
- Remove monkey patching (and private implementation introspection) of `Rails.logger`.
- Educate users on how to get previous behaviour if it's desired.

**Comment by ioquatix** [comment]
Just for reference, here is a monkey patch to bypass the problem entirely:

```ruby
class ActiveSupport::Logger
  def self.logger_outputs_to?(*)
    true
  end
end
```

**Comment by ioquatix** [comment]
Also for reference, the PR to introduce `logger_outputs_to?` https://github.com/rails/rails/pull/22933

**Comment by ioquatix** [comment]
cc @rafaelfranca @schneems perhaps you have some input on this.

**Comment by ioquatix** [comment]
This is still an issue in Rails 7.

**Comment by ioquatix** [comment]
@tenderlove what do you think of this? Can we get some movement on this issue? It makes logging harder than it should be.

**Comment by tenderlove** [comment]
> Default to Rails.logger = Logger.new(STDERR) (& don't log to log files).

I don't think this is going to happen.  Logging to files has been the default for so long, and it's not likely to change.  We should make `Rails.logger = Logger.new(STDERR)` work correctly though.

> I suggest some kind of tee logger should be appropriate, but I think it's better if we keep this simple in Rails and let logger gem solve this problem.

This seems good to me, but I don't have time to do the legwork.  If we have a tee logger available, then I'd be happy to change Rails to use it.

**Comment by ioquatix** [comment]
Okay, let me see if I can upstream some changes.

**Comment by ioquatix** [comment]
> I don't think this is going to happen. Logging to files has been the default for so long, and it's not likely to change. We should make Rails.logger = Logger.new(STDERR) work correctly though.

I recently ran into this issue - trying to dump CSV from a Rails app, and had it dump a bunch of logs directly in the middle of the CSV. I'm not sure why this is considered good behaviour.

**Comment by schneems** [comment]
> If we want to support output to multiple targets, this should be a feature of the log library not the interface for logging.

I made an issue that got nowhere: https://bugs.ruby-lang.org/issues/11955

Maybe you could have more luck?

> If we want to support output to multiple targets, this should be a feature of the log library not the interface for logging.

The reason I added that code was to avoid multiple logging IIRC.

Maybe having some kind of configuration to disable logging to stdout in IRB could be good, but the migration path should have some detection and warnings.

**Comment by ioquatix** [comment]
@schneems why do you log to STDOUT?

**Comment by skipkayhil** [comment]
> We should make `Rails.logger = Logger.new(STDERR)` work correctly though.

PR here #46309

---

I think there are multiple ideas being discussed so I wanted to be more explicit about them:

- Rails by default logs to files
  - this seems to have been addressed ("not likely to change")
- Rails ensures that _any_ logger is broadcast to console when running `rails server` or `rails console`
  - This is what I'm unclear on. How does using a tee logger prevent the need to check whether the configured logger is already broadcasting to the console? Or can this behavior be changed and only the _default_ logger needs to broadcast?
- Rails broadcasts to STDOUT for `rails server` and STDERR for `rails console`
  - It seems like there are a few places that fall back to a STDOUT logger but most fall back to a STDERR logger. Could the `server command` and e.g. https://github.com/rails/rails/blob/971fa17502deaf5d720a7bc0fdf8f258afb556b6/activejob/lib/active_job/logging.rb#L11 be made consistent and log to STDERR?

**Comment by ioquatix** [comment]
I agree with standardising logging to stderr and this would be a major step forward for logging in Rails.

> Rails ensures that any logger is broadcast to console when running rails server or rails console

I simply think that it should be a matter of how the user is interacting with the software rather than how it's run. If the output is to a tty (i.e. interactive), then it seems like Rails wants to log to both a terminal and a file.

So why not configure it like that?

```ruby
if $stderr.tty?
  config.logger = Logger::Tee.new($stderr, File.open(config.log_file))
else
  config.logger = Logger.new(File.open(config.log_file))
end
```

Alternatively, log to a pipe, and tee THAT to `$stderr` and a file.

**Comment by ioquatix** [comment]
https://github.com/rails/rails/pull/47143

We are now logging to stdout by default, and I wonder if we should remove the logging hack (since it's now the default). Additionally, I've proposed to @dhh we log to stderr instead of stdout.

**Comment by rafaelfranca** [comment]
We are only logging by default to stdout in production, not development, which is the reason why we have the broadcast feature.

**Comment by ioquatix** [comment]
That still creates mixed output, in production. i.e.

```
rake some:task:that:generates:csv > output.csv
```

In production, this will result in log messages being written to `output.csv`.

**Comment by rafaelfranca** [comment]
Right, but that is because stdout is the default, not because of the broadcast feature (what you call "hack"). I was answering if we can remove the broadcast feature, and right now we can't.

Probably we should broadcast to stderr instead of stdout.

**Comment by ioquatix** [comment]
> Probably we should broadcast to stderr instead of stdout.

Yes, this is a good step I the right direction.

**Comment by ioquatix** [comment]
I'm not calling the broadcast feature a hack, I'm referring to this piece of code:

https://github.com/rails/rails/blob/346ae79d5a11152f508df8d5506f8a0682ddb74b/activesupport/lib/active_support/logger.rb#L16-L20

Using `instance_variable_get` to dig into the internals is incredibly fragile.

**Comment by schneems** [comment]
I'm not happy with it, but it works and I didn't see any other alternative. I opened an issue on bugs.ruby-lang about opening an interface to expose the logger destination but it didn't go far. If you're interested in removing that hack that's probably the best next step.

**Comment by ioquatix** [comment]
@schneems I don't agree that the general logger interface should expose where it writes to because where a logger writes to can be very implementation specific, e.g. a logger might write to a log file AND stderr, or it might log to a log aggregation service over the network. Can you explain how those interfaces should work with your proposed "logger destination"?

**Comment by schneems** [comment]
I'm coming from a pragmatic place: Any code that solves the double logging problem while meeting all other specifications is fine.

If you don't like that interface, instead of exposing the destination perhaps more of a `respond_to` type pattern could be good where we directly ask the object it's capabilities. Something like:

```
logger.writes_to?(STDOUT)
```

Just for example. I'm really not attached to any particular interface though. 

At the end of the day, I think Rails just needs to answer the question: Will the end user see SQL logs on the console in development or not? 

- if no: inject a STDOUT logger
- if yes: do nothing, prevent double output to STDOUT

There might be some other way to do it, but for many releases Rails exposes the setting of the logger to Rails users. Spitballing here, but we could try some other gnarly solutions:

In a block re-assign $stdout to a StringIO instance, then intentionally trigger a log then inspect the StringIO instance and reset $stdout to see if it gets written. That's actually probably the most correct "can you do this thing" we can ask of the logger that doesn't violate its interface. However: it's gnarly. I don't know Rails would accept that code. I guess you could try to make that case if you want.

So it seems either we introduce a less-than-optimal API into Ruby to get rid of an awful hack in Rails, or we introduce a possibly questionable hack into Rails to avoid a different hack in Rails.

**Comment by ioquatix** [comment]
> At the end of the day, I think Rails just needs to answer the question: Will the end user see SQL logs on the console in development or not?

My position would be WHY do we need to answer that, why can't we just be declarative about it?

**Comment by schneems** [comment]
> If users choose to set a logger that outputs to stderr instead of stdout, why would we want to change that?

We don’t want to change the logger in either of those cases. Here’s the problem from the beginning:

We want to: log dev sql statements to the console even when a user is choosing to log to a file.

what my patch does: detects when someone is already logging to stdout so we can skip the magic logic that adds an extra log destination so people logging to files see an output in their development console. Without my patch, then someone using STDOUT as the output destination sees SQL statements in their console twice.

FWIW if we defaulted to logging to stdout (or stderr) then this problem goes away as it’s the assumed default and we don’t have to special case inject a logger based on magic conditions.

**Comment by ioquatix** [comment]
> We want to: log dev sql statements to the console even when a user is choosing to log to a file.

Okay sure, but why?

Why bypass what the user asks for?

> FWIW if we defaulted to logging to stdout (or stderr) then this problem goes away as it’s the assumed default and we don’t have to special case inject a logger based on magic conditions.

There were some changes recently: https://github.com/rails/rails/pull/47138 but it doesn't affect development environment.

**Comment by schneems** [comment]
> Okay sure, but why?

It's a feature of Rails. It helps you understand what's actually happening when you run your code. 

```
~ $ rails c
irb(main):001:0> User.where(github: "Tabby").first
D, [2023-02-07T21:42:18.624272 #7] DEBUG -- :   User Load (2.1ms)  SELECT "users".* FROM "users" WHERE "users"."github" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["github", "Tabby"], ["LIMIT", 1]]
=> nil
irb(main):003:0> User.where(email: "t<redacted>y@b<redacted>n.online").first
D, [2023-02-07T21:42:40.034092 #7] DEBUG -- :   User Load (2.0ms)  SELECT "users".* FROM "users" WHERE "users"."email" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["email", "t<redacted>y@b<redacted>n.online"], ["LIMIT", 1]]
=> nil
```

**Comment by ioquatix** [comment]
What about some variation of:

```
if $stdout.tty?
  config.logger = Logger.new($stdout)
else
  config.logger = Logger.new("logs/development.log")
end
```

**Comment by ioquatix** [comment]
I want to think about your question before writing a full answer, but in short:

> My ideal solution would be for Rails to log to either STDOUT or STDERR by default and make logging to files an opt-in case. But I think that is a much larger task.

I agree with this and think we should aim for this.
